digdag-core/src/main/java/io/digdag/core/workflow/WorkflowExecutor.java

```diff
@@ -482,46 +484,62 @@ public void runWhile(BooleanSupplier cond)
             throws InterruptedException
     {
         try (TaskQueuer queuer = new TaskQueuer()) {
-            Instant date = sm.getStoreTime();
-            propagateBlockedChildrenToReady();
-            retryRetryWaitingTasks();
-            enqueueReadyTasks(queuer);  // TODO enqueue all (not only first 100)
-            propagateAllPlannedToDone();
-            propagateSessionArchive();
-
-            //IncrementalStatusPropagator prop = new IncrementalStatusPropagator(date);  // TODO doesn't work yet
-            int waitMsec = INITIAL_INTERVAL;
-            while (cond.getAsBoolean()) {
-                //boolean inced = prop.run();
-                //boolean retried = retryRetryWaitingTasks();
-                //if (inced || retried) {
-                //    enqueueReadyTasks(queuer);
-                //    propagatorNotice = true;
-                //}
-
+            tm.begin(() -> {
+                Instant date = sm.getStoreTime();
                 propagateBlockedChildrenToReady();
                 retryRetryWaitingTasks();
-                enqueueReadyTasks(queuer);
-                boolean someDone = propagateAllPlannedToDone();
+                enqueueReadyTasks(queuer);  // TODO enqueue all (not only first 100)
+                propagateAllPlannedToDone();
+                propagateSessionArchive();
+                return null;
+            });
 
-                if (someDone) {
-                    propagateSessionArchive();
+            //IncrementalStatusPropagator prop = new IncrementalStatusPropagator(date);  // TODO doesn't work yet
+            final AtomicInteger waitMsec = new AtomicInteger(INITIAL_INTERVAL);
+            while (true) {
+                if (tm.<Boolean>begin(() -> !cond.getAsBoolean())) {
+                    break;
                 }
-                else {
+
+                tm.begin(() -> {
```

Comment:
I think here uses smaller transactions to avoid potential deadlocks.
Smaller transaction means almost one transaction per database operation.
For example,

* propagateBlockedChildrenToReady: one transaction for sm.findDirectParentsOfBlockedTasks, separated transactions for each sm.lockTaskIfExists calls.
* propagateAllPlannedToDone: one transaction for sm.findTasksByState, separated transactions for each sm.lockTaskIfExists calls.
* enqueueReadyTasks: one transaction per task